pandas/io/excel/_xlrd.py

```diff
@@ -120,9 +120,17 @@ def _parse_cell(cell_contents, cell_typ):
             elif cell_typ == XL_CELL_NUMBER:
                 # GH5394 - Excel 'numbers' are always floats
                 # it's a minimal perf hit and less surprising
-                val = int(cell_contents)
-                if val == cell_contents:
-                    cell_contents = val
+                try:
+                    val = int(cell_contents)
```

Comment:
Can you use `maybe_downcast_numeric`

Reply:
Do you think I should (is your comment a question or a request)?
I didn't know `maybe_downcast_numeric`, but I assume it could be used here. Without knowing `maybe_downcast_numeric` in detail, it appears to me that it likely will have a slight behavioural change compared to what the function does today (any which my PR retains, except for the exceptions). That said, the existing code definitely looks a bit "ad-hoc" to me, so it may be a good opportunity for a behavioural change. Let me know.

Reply:
More of a request. It appears the existing code is an ad-hoc way of downcasting float to int instead of maybe_downcast_numeric

Reply:
what is the slight behavioral change?

Reply:
I'm not *sure* there is one. Given the complexity of `maybe_downcast_numeric` compared to `int(cell_contents)` I was assuming there will be all sorts of corner cases where `maybe_downcast_numeric` will give a different result, not just the inf's and nan's. Anyway, I'll try to figure out how to use `maybe_downcast_numeric` here instead.

Reply:
Actually, it appears that `maybe_downcast_numeric` only applies to array-like objects, but `cell_contents` here is in fact a python `float`. Naive application of `maybe_downcast_numeric` causes an error, and wrapping the value into an array scalar just for the test feels a bit of an overkill here, no? Is there a similar function for scalars?

Reply:
Don't think so. I think `maybe_downcast_numeric(np.array([cell_contents]), np.int64)` should be fine

Reply:
`maybe_downcast_numeric(np.array([cell_contents]), np.dtype("int64"))` causes `FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison` downstream and `maybe_downcast_numeric(np.array(cell_contents), np.dtype("int64"))` causes `TypeError: unhashable type: 'numpy.ndarray'` downstream. `maybe_downcast_numeric(np.array(cell_contents), np.dtype("int64")).item()` works, should I use that?
Or would you prefer a more through refactor (which I won't be able to provide)?

Reply:
> maybe_downcast_numeric(np.array(cell_contents), np.dtype("int64")).item()

Yup this would be preferable

Reply:
... and, here we have that behaviour change:
```
self = <pandas.tests.io.excel.test_readers.TestReaders object at 0x2bebcb7d0>, read_ext = '.xls'

    def test_reader_converters(self, read_ext):
        basename = "test_converters"
    
        expected = DataFrame.from_dict(
            {
                "IntCol": [1, 2, -3, -1000, 0],
                "FloatCol": [12.5, np.nan, 18.3, 19.2, 0.000000005],
                "BoolCol": ["Found", "Found", "Found", "Not found", "Found"],
                "StrCol": ["1", np.nan, "3", "4", "5"],
            }
        )
    
        converters = {
            "IntCol": lambda x: int(x) if x != "" else -1000,
            "FloatCol": lambda x: 10 * x if x else np.nan,
            2: lambda x: "Found" if x != "" else "Not found",
            3: lambda x: str(x) if x else "",
        }
    
        # should read in correctly and set types of single cells (not array
        # dtypes)
        actual = pd.read_excel(
            basename + read_ext, sheet_name="Sheet1", converters=converters
        )
>       tm.assert_frame_equal(actual, expected)

/Users/yves/git-public/pandas/pandas/tests/io/excel/test_readers.py:509: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
testing.pyx:55: in pandas._libs.testing.assert_almost_equal ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   AssertionError: DataFrame.iloc[:, 1] (column name="FloatCol") are different
E   
E   DataFrame.iloc[:, 1] (column name="FloatCol") values are different (20.0 %)
E   [index]: [0, 1, 2, 3, 4]
E   [left]:  [12.5, nan, 18.3, 19.2, nan]
E   [right]: [12.5, nan, 18.3, 19.2, 5e-09]
E   At positional index 4, first diff: nan != 5e-09

testing.pyx:173: AssertionError
```

I believe I saw some code looking for small numbers in `maybe_downcast_numeric`. I have the impression that `maybe_downcast_numeric` serves a different purpose than the ad-hoc conversion we had here before. Note that there also seems to be a `maybe_downcast_numeric` somewhere further downstream from the code location we are looking at (within a `_infer_types` function call, which I only know because it appeared in the tracebacks of the other two variants of how to apply `maybe_downcast_numeric` here).

Given the behavioural change, I'd suggest to stick with the original fix without `maybe_downcast_numeric`, and leave such a change to either an actual bug, or a refactoring of the whole `read_excel` machinery. There may be a contract between the engine and the downstream machinery that the engine's only responsibility is to "get the data in", and the downstream machinery handles all "type guessing" and converting. That said, I also tried not converting to int at all; that also causes regressions in the tests.

Reply:
So to summarise: Both `maybe_downcast_numeric` and not downcasting at all cause behavioural changes compared to the original code (apparently introduced in #5394, 10 years ago). I tried, and each causes existing unit-tests to regress. Thus, I think such a change is out of scope for this bugfix, and I suggest to stick with the current state of the PR.

Reply:
Given that the input are pure-python floats exclusively, maybe using the functions from `math` may be more appropriate? Also, `isfinite` is equivalent and more performant to `not (isinf or isna)`, as long as we're dealing just  with floats (which we are). Thus, I pushed a change now using `math.isfinite`, let me know if you prefer `np.isfinite` or insist on your exact proposition.